CodingStandard.md

```diff
@@ -0,0 +1,1425 @@
+conventions.
+
+If a scenario is not covered on this page, use good judgment and remember
+*consistency is king*.  If there is an established way of doing something in
+other parts of the code base, that's probably the way it should be done.
+
+If the same question is likely to arise again later in the project, notify
+the Lead Software Engineer, so he or she can update this page for everyone's
+benefit.
+
+Note that the use of constructs or mechanisms for which justification is
+required is not discouraged. Requiring justification in these cases is simply
+intended to prevent misuse and to capture rationale for future reference.
+
+
+## General Standards
+
+**Language Specification**
+
+All production code should be written in the C++ programming language and should
+be compliant with the C++14 (ISO/IEC 14882:2014) language standard, except where
+such standards are not supported by the Robot Operating System (ROS) framework.
+If such instances are found, notify the Software Director so that they may be
+included in this document.
+
+**Coding Standard**
+
+Project standards and guidelines defined in this document should be adhered to
+for all production code. This document shall be maintained and updated as those
+standards are modified or replaced.
+
+**Names**
+
+Names (types, variables, functions, constants, namespaces, enumerators,
+macros, files, etc.) must be descriptive of their purpose and consistent.
+
+**Comments**
+
+Comments must be used to clarify parts of code that merit explanation,
+must be intelligible, and must be maintained.
+
+*[Question: Should we add a standard here for including Doxygen as a supported 
+
+tool?]*
+
+**Compilation**
+
+Code should compile and link with no errors or warnings at the `-Wall` `-Wextra`
+warning level.  Unavoidable warnings may be disabled locally, in as small a
+section of code as possible, with rationale stated in a comment wherever the
+warning is disabled.
+
+
+**#Includes**
+
+System-wide include files (e.g. stdafx.h) must contain a minimal amount of
+definitions.
+
+#include directives in a file must only be preceded by other preprocessor
+directives or comments. Exception: files that contain large data tables
+(e.g. a CRC coefficient lookup table) may be included at any location to
+promote readability.
+
+
+**Header Files**
+
+Every header file should be able to be compiled in isolation, meaning it
+includes any headers that its contents depend upon.  Relying on transitive
+#includes of other headers can make maintenance difficult and introduce
+subtle issues.
+
+
+**Memory**
+
+A function must not return a pointer or reference to memory allocated
+within that function with temporary or automatic storage duration.
+
+Classes that allocate memory with dynamic storage duration, including
+any memory allocated with the new operator, or allocate other shared resources
+must explicitly define copy constructors and copy assignment operators.
```

Comment:
Copy assignment for shared resource? why not shared_ptr?

Reply:
or even better, not have shared resources.

Reply:
You're right---shared_ptr would be preferred. Will update.